tests/scripts/all.py

```diff
@@ -0,0 +1,51 @@
+#!/usr/bin/env python3
+
+import subprocess
+import sys
+from types import FunctionType
+
+BASH_COMPONENT_CODE = '''
+global component_{0}
+def component_{0}():
+    component_proc = subprocess.Popen(['tests/scripts/all.sh', '{0}'], stdout=subprocess.PIPE)
+    for line in component_proc.stdout:
+        line = str(line, 'utf-8')
+        print(line, end='')
+'''
+
+def import_bash_components_file(components_file):
+    # Get a list of functions defined in the components file
+    get_fns_cmd = ['/bin/bash', '-c',
+                   'source {} && compgen -A function'.format(components_file)]
+
+    function_list = subprocess.check_output(get_fns_cmd)
+    function_list = str(function_list, 'utf-8').split()
+
+    # Get the components
+    component_names = filter(lambda x: x.startswith('component_'), function_list)
+    # Remove 'component_' prefix
+    component_names = list(map(lambda x: x[len('component_'):], component_names))
+
+    for component in component_names:
+        # Create a new python function that runs the component in all.sh
+        exec(BASH_COMPONENT_CODE.format(component))
+
+
+import_bash_components_file('tests/scripts/components.sh')
+
+# Get all functions starting with 'component_'
+components = dict(filter(lambda x: x[0].startswith('component_'), locals().items()))
+
+if len(sys.argv) > 1:
```

Comment:
What's the goal for the transition? What do you want to be able to do in the short term, that you can't do (easily) with the shell code? What would be the entry point during the transition?

`all.sh` has many command line options. All of them were thought to be useful at some point. A few of them may be obsolete, but most are still useful. If the entry point is `all.py`, it needs to support all of the useful options.

If you want to be able to write some python components because you want to do things that are complicated in bash, then `all.py` doesn't need all these command line options. But then it doesn't need to call `all.sh`: `all.sh` would do the calling.

Reply:
Switching `all.sh` to python is a long-term goal for me, but I might not have the same objectives that you have. For me, the goals are:

* More robust error recovery. Error handling in bash is subtle and there's a risk that some errors are ignored.
* Easier to maintain, especially for people who aren't bash experts.
* Easier to write certain “structured” components. For example, we could meld `depends.py` and `test-ref-configs.pl` into the python test runner and then request a subset of those.
* Easier to write parametrized components, and modify them to run something different from what the CI does. For example, take a component and change it from gcc to clang in a simple way.
* Better automation for certain things that come up often, for example creating a temporary build directory or whether to start with generated files.
* Make it easy to run part of a component. For example `tests/scripts/run.py --config test_foo` to only run `config.py` but not `make`, prior to running `make` with debug flags for local testing.
* Support some component analysis, for example “list the components whose configuration has AES disabled”.
* Support grouping of components, for example “run all the components that have SSL tests”.

For that, my plan for the transition would be:

* Create a python driver, starting out as an empty shell.
* Gradually add the ability to run certain components, each time adding support for related driver features. For example:
    * Add the equivalent of `support_xxx` the first time we add a component with a system dependency.
    * Add `--openssl-…` and such options the first time we add a component that runs SSL tests.
    * Add `--no-armcc` when we add an armcc component.
* Once the python driver has all the capabilities of the bash driver, make it the primary entry point, calling `all.sh` for individual components that aren't yet converted to python.
* Convert more components to python.
* For the remaining components that don't have any particular structure that's useful to translate, just include them as shell code embedded in the python code. Remove the `all.sh` driver.